packages/headless/src/app/engine.ts

```diff
@@ -168,6 +168,9 @@ function buildCoreEngine<
 
   return {
     addReducers(reducers: ReducersMapObject) {
+      if (reducerManager.contains(reducers)) {
+        return;
+      }
```

Comment:
Posting here for better threads.

**What reducer case is invoked?**
>Whenever replaceReducers is executed, it re-execute (all?) reducers already present, including restoreSearchParameters. So, it would get executed something like .. 10+ times on each initial load (for each controller getting instantiated).

When all reducers are re-executed, I'm curious which reducer _case_ is invoked? In the blurb above, it says the `restoreSearchParameters` _action_ is re-executed? So does it mean `replaceReducers` re-invokes the last executed action? If it was a different action, that action case would be re-invoked?

**Hmm, strange**
> The payload is different from the "original" one, and instead of receiving the full array of selection from the URL, we receive a payload that is incomplete

> The reason why the payload is incomplete is still a mystery to me

Yeah, this is strange. If the last action is re-invoked, I'd expect the payload to be either identical, or empty, not something in between?

**Could the issue reappear with an unregistered reducer?**
> This means the restoreSearch reducer only gets executed once on load.

Just to clarify this statement, the `restoreSearchParameters` is an action that is handled by multiple reducers.

The change made by this PR would prevent re-invoking the last action (current hypothesis) if the same reducer was re-registered. But it would not prevent re-invoking if a different reducer was registered. Could this cause the issue reappear?


**Performance**
> This is also solving a relatively big performance issue

Not ruling this out, but I'm going to put aside the performance aspect for a moment (or was there a measurable improvement?). My thinking is all reducers are re-invoked all the time (on every action dispatch), but the computation is so small, it happens quickly? (it's mostly rejected switch case statements?)

Reply:
> So does it mean replaceReducers re-invokes the last executed action? If it was a different action, that action case would be re-invoked?

It seems it re-execute all actions that were dispatched in that state change "loop", until it has a chance to "commit" to state. "Commit" being a redux concept/internal logic.

> But it would not prevent re-invoking if a different reducer was registered. Could this cause the issue reappear?

Oh yes, it could ! Any result specific controller (or component) that would need a reducer not already present could cause a problem. InteractiveResult only needs configuration (which is already present), so it's fine.

This is not an ideal/perfect fix !

> computation is so small, it happens quickly? (it's mostly rejected switch case statements?)

Yes, in the grand scheme of things, yes just doing "pure state" manipulation is always going to be 100x time faster than painting anything in the DOM :D 

Still, I'm not sure why we would not want to add such a relatively simple if check that verify if the reducer is already present, instead of just replacing it with "itself" for no reason. 

Is there a reason I'm unaware that would benefit from re-running the same function dozen of times on init ?

Reply:
The redux commit is new knowledge, I wasn't aware that redux was using a batching mechanism. It's wishful thinking, but I see the project is on `redux@4.0.5` from 2 years ago. Maybe the the latest version might contain a fix in this mechanism? There are several closed issues related to `replaceReducer` [here](https://github.com/reduxjs/redux/issues?q=is%3Aissue+replaceReducer+).

Since this is not a perfect fix, I'm worried we'll see the issue down the line, in a scenario that might be harder to debug.
In seems that an alternate solution could have been to remove the call to `addReducers` from the controller?

On adding the `if` check, I'm thinking about [this line too](https://github.com/coveo/ui-kit/pull/1146/files#diff-bff128bba028878c567bc8c549c38e010cc55d1982e9f44c94d019f52bdfa0beR21). Today, we do not allow people to override reducers, so the new `if` check does not break existing implementations.

However, if one day we want to allow it, we would need to remove the linked line along with the change in this PR. That will uncover the bug. I am trying to focus on the core issue instead of the side-benefits of the current approach,

Reply:
After more investigation, this seems to be a problem with redux dev tool actually: https://github.com/zalmoxisus/redux-devtools-extension/issues/326

I'll try to see if there's another solution (potentially configuring dev tools in a different manner, or upgrading)